File: windows/CursorTokenCore/CursorAccountPaste.cs
using System.Text.RegularExpressions;

namespace CursorTokenCore;

public sealed class AccountPasteItem
{
    public string Kind { get; init; } = "";
    public string Token { get; init; } = "";
    public string Email { get; init; } = "";
    public string Password { get; init; } = "";
    public string Message { get; init; } = "";
}

public static class CursorAccountPaste
{
    const string EmailToken = @"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}";
    static readonly Regex LabeledBoth = new(
        @"^(?:账号|帐号|账户|邮箱|用户名)\s*[：:]\s*(" + EmailToken + @")\s*密码\s*[：:]\s*(.+)$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);
    static readonly Regex LabeledEmail = new(
        @"^(?:账号|帐号|账户|邮箱|用户名)\s*[：:]\s*(" + EmailToken + @")\s*$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);
    static readonly Regex LabeledPassword = new(
        @"^密码\s*[：:]\s*(.+)$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);
    static readonly Regex EmailAtStart = new(
        @"^(" + EmailToken + @")(.*)$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);

    public static bool LooksLikeToken(string? line)
    {
        var text = (line ?? "").Trim();
        if (text.Length == 0) return false;
        var lower = text.ToLowerInvariant();
        if (lower.Contains("workoscursorsessiontoken=")) return true;
        if (lower.Contains("%3a%3a") || text.Contains("::")) return true;
        var parts = text.Split('.');
        return parts.Length == 3 && parts.All(p => p.Length > 0);
    }

    public static bool IsSingleToken(string? text)
    {
        var items = Parse(text);
        return items.Count == 1 && items[0].Kind == "token";
    }

    public static List<AccountPasteItem> Parse(string? text)
    {
        var raw = (text ?? "").Replace("\r\n", "\n").Replace('\r', '\n');
        var items = new List<AccountPasteItem>();
        var pendingEmail = "";

        void FlushPending()
        {
            if (pendingEmail.Length == 0) return;
            items.Add(Item("error", email: pendingEmail, message: "只有账号没有密码"));
            pendingEmail = "";
        }

        foreach (var rawLine in raw.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0) continue;

            if (TryLabeledBoth(line, out var bothEmail, out var bothPassword))
            {
                FlushPending();
                items.Add(Item("credentials", email: bothEmail, password: bothPassword));
                continue;
            }

            var emailOnly = ParseLabeledEmail(line);
            if (emailOnly.Length > 0)
            {
                FlushPending();
                pendingEmail = emailOnly;
                continue;
            }

            if (TryLabeledPassword(line, out var passwordOnly))
            {
                if (pendingEmail.Length > 0 && passwordOnly.Length > 0)
                {
                    items.Add(Item("credentials", email: pendingEmail, password: passwordOnly));
                    pendingEmail = "";
                }
                else if (pendingEmail.Length > 0)
                    FlushPending();
                else
                    items.Add(Item("error", message: "只有密码没有账号"));
                continue;
            }

            if (TryEmailSeparator(line, out var sepEmail, out var sepPassword))
            {
                FlushPending();
                items.Add(Item("credentials", email: sepEmail, password: sepPassword));
                continue;
            }

            if (LooksLikeToken(line))
            {
                FlushPending();
                items.Add(Item("token", token: line));
                continue;
            }

            FlushPending();
            items.Add(Item("error", message: "无法识别"));
        }

        FlushPending();
        return items;
    }

    static AccountPasteItem Item(string kind, string token = "", string email = "", string password = "", string message = "") =>
        new() { Kind = kind, Token = token, Email = email, Password = password, Message = message };

    static bool TryLabeledBoth(string line, out string email, out string password)
    {
        email = "";
        password = "";
        var match = LabeledBoth.Match(line);
        if (!match.Success) return false;
        email = CursorPasswordLogin.SanitizeEmail(match.Groups[1].Value);
        password = match.Groups[2].Value.Trim();
        return email.Length > 0 && password.Length > 0;
    }

    static string ParseLabeledEmail(string line)
    {
        var match = LabeledEmail.Match(line);
        return match.Success ? CursorPasswordLogin.SanitizeEmail(match.Groups[1].Value) : "";
    }

    static bool TryLabeledPassword(string line, out string password)
    {
        password = "";
        var match = LabeledPassword.Match(line);
        if (!match.Success) return false;
        password = match.Groups[1].Value.Trim();
        return true;
    }

    static bool TryEmailSeparator(string line, out string email, out string password)
    {
        email = "";
        password = "";
        var match = EmailAtStart.Match(line);
        if (!match.Success) return false;
        email = CursorPasswordLogin.SanitizeEmail(match.Groups[1].Value);
        if (email.Length == 0) return false;
        var rest = match.Groups[2].Value;
        if (rest.Length == 0) return false;
        var stripped = rest.TrimStart();
        if (stripped.StartsWith("----", StringComparison.Ordinal))
            password = stripped[4..].Trim();
        else if (stripped.StartsWith(':') || stripped.StartsWith('：'))
            password = stripped[1..].Trim();
        else if (rest[0] is ' ' or '\t')
            password = rest.Trim();
        else
            return false;
        return password.Length > 0;
    }
}

File: macos/Sources/CursorTokenCore/CursorAccountPaste.swift
import Foundation

public struct AccountPasteItem: Equatable {
    public let kind: String
    public let token: String
    public let email: String
    public let password: String
    public let message: String

    public init(kind: String, token: String = "", email: String = "", password: String = "", message: String = "") {
        self.kind = kind
        self.token = token
        self.email = email
        self.password = password
        self.message = message
    }
}

public enum CursorAccountPaste {
    static let emailToken = #"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}"#
    static let labeledBoth = regex(#"^(?:账号|帐号|账户|邮箱|用户名)\s*[：:]\s*("# + emailToken + #")\s*密码\s*[：:]\s*(.+)$"#)
    static let labeledEmail = regex(#"^(?:账号|帐号|账户|邮箱|用户名)\s*[：:]\s*("# + emailToken + #")\s*$"#)
    static let labeledPassword = regex(#"^密码\s*[：:]\s*(.+)$"#)
    static let emailAtStart = regex(#"^("# + emailToken + #")(.*)$"#)

    public static func looksLikeToken(_ line: String?) -> Bool {
        let text = (line ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty { return false }
        let lower = text.lowercased()
        if lower.contains("workoscursorsessiontoken=") { return true }
        if lower.contains("%3a%3a") || text.contains("::") { return true }
        let parts = text.split(separator: ".", omittingEmptySubsequences: false)
        return parts.count == 3 && parts.allSatisfy { !$0.isEmpty }
    }

    public static func isSingleToken(_ text: String?) -> Bool {
        let items = parse(text)
        return items.count == 1 && items[0].kind == "token"
    }

    public static func tokenValues(_ text: String?) -> [String] {
        let items = parse(text)
        guard !items.isEmpty, items.allSatisfy({ $0.kind == "token" }) else { return [] }
        return items.map(\.token).filter { !$0.isEmpty }
    }

    public static func parse(_ text: String?) -> [AccountPasteItem] {
        let raw = (text ?? "").replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
        var items: [AccountPasteItem] = []
        var pendingEmail = ""

        func flushPending() {
            guard !pendingEmail.isEmpty else { return }
            items.append(AccountPasteItem(kind: "error", email: pendingEmail, message: "只有账号没有密码"))
            pendingEmail = ""
        }

        for rawLine in raw.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            if line.isEmpty { continue }

            if let both = parseLabeledBoth(line) {
                flushPending()
                items.append(AccountPasteItem(kind: "credentials", email: both.0, password: both.1))
                continue
            }

            let emailOnly = parseLabeledEmail(line)
            if !emailOnly.isEmpty {
                flushPending()
                pendingEmail = emailOnly
                continue
            }

            if let passwordOnly = parseLabeledPassword(line) {
                if !pendingEmail.isEmpty && !passwordOnly.isEmpty {
                    items.append(AccountPasteItem(kind: "credentials", email: pendingEmail, password: passwordOnly))
                    pendingEmail = ""
                } else if !pendingEmail.isEmpty {
                    flushPending()
                } else {
                    items.append(AccountPasteItem(kind: "error", message: "只有密码没有账号"))
                }
                continue
            }

            if let separated = parseEmailSeparator(line) {
                flushPending()
                items.append(AccountPasteItem(kind: "credentials", email: separated.0, password: separated.1))
                continue
            }

            if looksLikeToken(line) {
                flushPending()
                items.append(AccountPasteItem(kind: "token", token: line))
                continue
            }

            flushPending()
            items.append(AccountPasteItem(kind: "error", message: "无法识别"))
        }

        flushPending()
        return items
    }

    static func regex(_ pattern: String) -> NSRegularExpression {
        try! NSRegularExpression(pattern: pattern)
    }

    static func firstMatch(_ re: NSRegularExpression, _ text: String) -> NSTextCheckingResult? {
        re.firstMatch(in: text, range: NSRange(text.startIndex..., in: text))
    }

    static func group(_ match: NSTextCheckingResult, _ index: Int, in text: String) -> String {
        guard let range = Range(match.range(at: index), in: text) else { return "" }
        return String(text[range])
    }

    static func parseLabeledBoth(_ line: String) -> (String, String)? {
        guard let match = firstMatch(labeledBoth, line) else { return nil }
        let email = CursorPasswordLogin.sanitizeEmail(group(match, 1, in: line))
        let password = group(match, 2, in: line).trimmingCharacters(in: .whitespacesAndNewlines)
        if email.isEmpty || password.isEmpty { return nil }
        return (email, password)
    }

    static func parseLabeledEmail(_ line: String) -> String {
        guard let match = firstMatch(labeledEmail, line) else { return "" }
        return CursorPasswordLogin.sanitizeEmail(group(match, 1, in: line))
    }

    static func parseLabeledPassword(_ line: String) -> String? {
        guard let match = firstMatch(labeledPassword, line) else { return nil }
        return group(match, 1, in: line).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func parseEmailSeparator(_ line: String) -> (String, String)? {
        guard let match = firstMatch(emailAtStart, line) else { return nil }
        let email = CursorPasswordLogin.sanitizeEmail(group(match, 1, in: line))
        if email.isEmpty { return nil }
        let rest = group(match, 2, in: line)
        if rest.isEmpty { return nil }
        let stripped = rest.drop { $0 == " " || $0 == "\t" }
        let password: String
        if stripped.hasPrefix("----") {
            password = String(stripped.dropFirst(4)).trimmingCharacters(in: .whitespacesAndNewlines)
        } else if stripped.first == ":" || stripped.first == "：" {
            password = String(stripped.dropFirst()).trimmingCharacters(in: .whitespacesAndNewlines)
        } else if rest.first == " " || rest.first == "\t" {
            password = rest.trimmingCharacters(in: .whitespacesAndNewlines)
        } else {
            return nil
        }
        if password.isEmpty { return nil }
        return (email, password)
    }
}
